src/celeritas/global/Stepper.cc

```diff
@@ -0,0 +1,155 @@
+
+    // TODO: add accessors instead of looking these up as strings
+    const ActionManager& action_mgr = *core.action_mgr();
+    for (const char* action_label : {
+             "pre-step",
+             "along-step",
+             "physics-discrete-select",
+             "geo-boundary",
+         })
+    {
+        actions.push_back(action_mgr.find_action(action_label));
+        CELER_VALIDATE(actions.back(),
+                       << "missing action '" << action_label << "'");
+    }
+
+    for (ActionId aid : core.physics()->model_actions())
+    {
+        CELER_ASSERT(aid);
+        actions.push_back(aid);
+    }
+
+    return actions;
+}
+//---------------------------------------------------------------------------//
+} // namespace
+
+//---------------------------------------------------------------------------//
+// Construct with problem parameters and setup options
+template<MemSpace M>
+Stepper<M>::Stepper(Input input)
+    : params_(std::move(input.params))
+    , num_initializers_(input.num_initializers)
+{
+    CELER_EXPECT(params_);
+    CELER_VALIDATE(input.num_track_slots > 0,
+                   << "number of track slots has not been set");
+    states_ = CollectionStateStore<CoreStateData, M>(*params_,
+                                                     input.num_track_slots);
+
+    // Construct main actions
+    actions_ = build_action_vec(*params_);
+
+    if (input.post_step_callback)
+    {
+        actions_.push_back(input.post_step_callback);
+    }
+
+    core_ref_.params = get_ref<M>(*params_);
+    core_ref_.states = states_.ref();
+
+    CELER_ENSURE(!actions_.empty());
+}
+
+//---------------------------------------------------------------------------//
+// Default destructor
+template<MemSpace M>
+Stepper<M>::~Stepper() = default;
+
+//---------------------------------------------------------------------------//
+// Transport existing states
+template<MemSpace M>
+auto Stepper<M>::operator()() -> result_type
+{
+    CELER_EXPECT(*this);
+
+    // TODO: refactor initialization
+    CELER_VALIDATE(inits_, << "no primaries were given");
+
+    result_type result;
+
+    // Create new tracks from queued primaries or secondaries
+    initialize_tracks(core_ref_, &inits_);
+    result.active = states_.size() - inits_.vacancies.size();
+
+    // Launch the interaction kernels for all applicable models
+    const ActionManager& action_mgr = *params_->action_mgr();
+    for (ActionId action : actions_)
+    {
+        action_mgr.invoke<M>(action, core_ref_);
+    }
+
+    // Create track initializers from surviving secondaries
+    extend_from_secondaries(core_ref_, &inits_);
+
+    // Get the number of track initializers and active tracks
+    result.alive  = states_.size() - inits_.vacancies.size();
+    result.queued = inits_.initializers.size();
+
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+// Transport existing states and these new primaries
```

Comment:
The function does call "step initialized tracks" after initializing the new ones but I'll improve the documentation (especially since I forgot to update the doxygen docs).

Reply:
Ah you're right, missed that at the bottom!